src/common/components/multiSelectDropdown/MultiSelectDropdown.tsx

```diff
@@ -17,93 +17,118 @@ export type Option = {
   const [internalInput, setInternalInput] = React.useState('');
   const input = inputValue !== undefined ? inputValue : internalInput;
 
-  const inputWrapper = React.useRef<HTMLDivElement | null>(null);
+  const dropdown = React.useRef<HTMLDivElement | null>(null);
+  const toggleButton = React.useRef<HTMLButtonElement | null>(null);
   const inputRef = React.useRef<HTMLInputElement | null>(null);
-  const clearButtonRef = React.useRef<HTMLButtonElement | null>(null);
 
   const filteredOptions = React.useMemo(() => {
     return [
-      showSelectAll
-        ? {
-            text: selectAllText || t('commons.multiSelectDropdown.selectAll'),
-            value: SELECT_ALL,
-          }
-        : undefined,
-      ...options.filter(option =>
+      showSelectAll && {
+        text: selectAllText || t('commons.multiSelectDropdown.selectAll'),
+        value: SELECT_ALL,
+      },
+      ...options.filter((option) =>
         option.text.toLowerCase().includes(input.toLowerCase())
       ),
-    ].filter(e => e) as Option[];
+    ].filter((e) => e) as Option[];
   }, [input, options, selectAllText, showSelectAll, t]);
 
-  const handleInputValueChange = (val: string) => {
-    setInternalInput(val);
-
-    if (setInputValue) {
-      setInputValue(val);
-    }
-  };
+  const handleInputValueChange = React.useCallback(
+    (val: string) => {
+      setInternalInput(val);
 
-  const dropdown = React.useRef<HTMLDivElement | null>(null);
+      if (setInputValue) {
+        setInputValue(val);
+      }
+    },
+    [setInputValue]
+  );
 
-  const [
+  const {
     focusedIndex,
-    setupKeyboardNav,
-    teardownKeyboardNav,
-  ] = useKeyboardNavigation({
+    setup: setupKeyboardNav,
+    teardown: teardownKeyboardNav,
+  } = useKeyboardNavigation({
     container: dropdown,
     listLength: filteredOptions.length,
+    onKeyDown: (event: KeyboardEvent) => {
+      // Handle keyboard events only if current element is focused
+      if (!isComponentFocused()) return;
+
+      switch (event.key) {
+        // Close menu on ESC key
+        case 'Escape':
+          setIsMenuOpen(false);
+          setFocusToToggleButton();
+          break;
+        case 'ArrowUp':
+          ensureDropdownIsOpen();
+          break;
+        case 'ArrowDown':
+          ensureDropdownIsOpen();
+          break;
+        case 'Enter':
+          if (isToggleButtonFocused()) {
+            handleToggleButtonClick();
+          }
+          event.preventDefault();
+      }
+    },
   });
   const [isMenuOpen, setIsMenuOpen] = React.useState(false);
 
-  const isComponentFocused = React.useCallback(() => {
+  const isComponentFocused = () => {
     const active = document.activeElement;
-    const current = dropdown && dropdown.current;
+    const current = dropdown.current;
 
-    if (current && active instanceof Node && current.contains(active)) {
-      return true;
-    }
-    return false;
-  }, [dropdown]);
+    return !!current?.contains(active);
```

Comment:
.contains returns boolean? Is "!!" Redundant?

Reply:
If current is not defined it return undefined because contains is not called.